### PR TITLE
pythonPackages.pysideShiboken: fix build with python 3.9

### PR DESCRIPTION
**Pull request by jnetod**
###### Motivation for this change

ZHF: #122042

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 123778` at 24d76dc6 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>natron</li>
  </ul>
</details>
<details>
  <summary>1 package skipped due to time constraints:</summary>
  <ul>
    <li>fmbt</li>
  </ul>
</details>
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li>python38Packages.pyside</li>
    <li>python38Packages.pysideShiboken</li>
    <li>python38Packages.pysideTools</li>
    <li>python39Packages.pyside</li>
    <li>python39Packages.pysideShiboken</li>
    <li>python39Packages.pysideTools</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/python-modules/pyside/shiboken.nix:67:5:
```
   |
67 |     maintainers = [ ];
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md">stale-substitute</a>

Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '"${GENERATORRUNNER_PLUGIN_DIR}"' doesn't match anything in file 'generator/CMakeLists.txt'
Near pkgs/development/python-modules/pyside/shiboken.nix:64:0:
```
   |
64 |     description = "Plugin (front-end) for pyside-generatorrunner, that generates bindings for C++ libraries using CPython source code";
   | ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/development/python-modules/pyside/shiboken.nix:65:5:
```
   |
65 |     license = licenses.gpl2;
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/development/python-modules/pyside/shiboken.nix:64:0:
```
   |
64 |     description = "Plugin (front-end) for pyside-generatorrunner, that generates bindings for C++ libraries using CPython source code";
   | ^
```
</li>
  </ul>
</details>

File: pkgs/development/python-modules/pyside/shiboken.nix [hunk 1]
```diff
@@ -39,8 +42,17 @@ buildPythonPackage rec {
       \"$\{GENERATORRUNNER_PLUGIN_DIR}\" lib/generatorrunner/
   '';
 
-  # gcc6 patch was also sent upstream: https://github.com/pyside/Shiboken/pull/86
-  patches = [ ./gcc6.patch ] ++ (lib.optional (isPy35 || isPy36 || isPy37) ./shiboken_py35.patch);
+  patches = [
+    # gcc6 patch was also sent upstream: https://github.com/pyside/Shiboken/pull/86
+    ./gcc6.patch
+    (lib.optional (isPy35 || isPy36 || isPy37 || isPy38 || isPy39 ) ./shiboken_py35.patch)
```
Comment (jonringer) [inline pkgs/development/python-modules/pyside/shiboken.nix:48]:
```suggestion
    (lib.optional (pythonAtLeast "3.5") ./shiboken_py35.patch)
```

File: pkgs/development/python-modules/pyside/shiboken.nix [hunk 2]
```diff
@@ -39,8 +42,17 @@ buildPythonPackage rec {
...
+    (fetchurl {
+      # https://github.com/pyside/Shiboken/pull/90
```
Comment (jonringer) [inline pkgs/development/python-modules/pyside/shiboken.nix:50]: do you mind also commenting why this is relevant? easier when revisiting this

**Comment by jonringer** [comment]
otherwise LGTM

```
https://github.com/NixOS/nixpkgs/pull/123778

1 package marked as broken and skipped:
natron

8 packages built:
fmbt python38Packages.pyside python38Packages.pysideShiboken python38Packages.pysideTools python39Packages.pyside python39Packages.pysideShiboken python39Packages.pysideTools tests.trivial
```

**Comment by risicle** [comment]
`nixpkgs-review` happy for me nixos x86_64

File: pkgs/development/python-modules/pyside/shiboken.nix (hunk 2 above)
Comment (jnetod) [reply pkgs/development/python-modules/pyside/shiboken.nix:50]: `tp_print` was mostly used on python 2.x, unused on python 3.x and was [removed on python 3.9](https://docs.python.org/3/whatsnew/3.9.html#id3) - that's why pysideShiboken broke.

The patch just add a check to ignore `tp_print` on python 3.x.

File: pkgs/development/python-modules/pyside/shiboken.nix (hunk 1 above)
Comment (jnetod) [reply pkgs/development/python-modules/pyside/shiboken.nix]: Done

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 123778 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>1 package marked as broken and skipped:</summary>  <ul>    <li>natron</li>  </ul></details><details>  <summary>8 packages built:</summary>  <ul>    <li>fmbt</li>    <li>python38Packages.pyside</li>    <li>python38Packages.pysideShiboken</li>    <li>python38Packages.pysideTools</li>    <li>python39Packages.pyside</li>    <li>python39Packages.pysideShiboken</li>    <li>python39Packages.pysideTools</li>    <li>tests.trivial</li>  </ul></details>

**Pull request merged**